docs/2023/puzzles/day14.md

```diff
@@ -2,10 +2,337 @@ import Solver from "../../../../../website/src/components/Solver.js"
+end rollUp
+```
+
+So, `rollUp` will always roll the stones up the platform - but the "up" is relative to the rotation we've set! So, if we set the rotation to `Direction.W`, the stones will roll to the left when we invoke `rollUp` - although in the current view of the model, it will look like they roll up. If we set the rotation to `Direction.S`, the stones will roll down. And so on. E.g., to roll left:
+
+```scala
+model.rotation = Direction.W
+rollUp(model)
+```
+
+This is achieved by the `cellsIterator` method which, under the hood, uses `model.extentX`, `model.extentY` and `model(x, y)` - the dimensions and the model accessor that are relative to the model rotation.
+
+We can also calculate the model loading as follows:
+
+```scala
+def totalLoading(model: Model): Int =
+  model.rotation = Direction.N
+  var loading = 0
+  for (_, y, c) <- cellsIterator(model) do c match
+    case 'O' => loading += model.extentY - y
+    case _ =>
+  loading
+```
+
+Since the model loading needs to always be calculated against the northern side of the model, we always set the rotation to `Direction.N` before calculating the loading.
+
+### Naive cycling
+
+The challenge requires us to cycle through the North-West-South-East tilts for a billion cycles. Let's implement the simplest possible cycling function:
+
+```scala
+def cycle(model: Model, times: Int): Unit =
+  for i <- 1 to times do
+    for cse <- Direction.values do
+      model.rotation = cse
+      rollUp(model)
+  model.rotation = Direction.N
+```
+
+This simple loop repeats a required number of times, and for each loop, we have a nested loop that iterates over all the directions (in order) and performs the rollup for each of those directions. At the end, the rotation is reset to North.
+
+### Performant cycling
+
+This solution looks good but won't get to the billion cycles in a reasonable time. The rollups take way too long.
+
+To optimize, we will use the intent of the operation: we cycle for so many times to eventually converge to a certain state. As the challenge puts it, "This process [cycling repeatedly] should work if you leave it running long enough, but you're still worried about the north support beams...".
+
+So, there's a possibility that the desired state will be achieved earlier than the billion cycles. Looks like a good case for a dynamic programming approach: we need to remember the states of the model we've seen before and what they look like after one cycle. And if the current state is in our cache, no need to compute the next state again.
+
+Furthermore, since the transitions between states are deterministic (a state A always leads to state B), the moment one state in our cache is followed by another, previously encountered, state from that cache, we can stop cycling and just calculate the final state from the cache.
+
+```scala
+import scala.util.boundary, boundary.break
+
+def cycle(model: Model, times: Int): Unit =
+  val chain = collection.mutable.ListBuffer.empty[String]
```

Comment:
If the aim here is all about optimisation shouldn't we not have a linear lookup?